tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/ShadowSystem.java

```diff
@@ -0,0 +1,137 @@
+/*******************************************************************************
+ * Copyright 2011 See AUTHORS file.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+package com.badlogic.gdx.tests.g3d.shadows.system;
+
+import java.util.Set;
+
+import com.badlogic.gdx.graphics.Camera;
+import com.badlogic.gdx.graphics.Cubemap.CubemapSide;
+import com.badlogic.gdx.graphics.g3d.environment.DirectionalLight;
+import com.badlogic.gdx.graphics.g3d.environment.PointLight;
+import com.badlogic.gdx.graphics.g3d.environment.SpotLight;
+import com.badlogic.gdx.graphics.g3d.utils.ShaderProvider;
+
+/** Shadow system provides functionalities to render shadows.
+ * <p>
+ * Typical use: <br />
+ *
+ * // Init system: <br />
+ * Array<ModelBatch> passBatches = new Array<ModelBatch>();<br />
+ * ModelBatch mainBatch; <br />
+ * ShadowSystem system = new XXXShadowSystem(cam, instances); <br />
+ * for (int i = 0; i < system.getPassQuantity(); i++) { <br />
+ * passBatches.add(new ModelBatch(system.getPassShaderProvider(i))); <br />
+ * } <br />
+ * mainBatch = new ModelBatch(system.getShaderProvider()) <br />
+ *
+ * // Render scene with shadows: <br />
+ * system.update(); <br />
+ * for (int i = 0; i < system.getPassQuantity(); i++) { <br />
+ * system.begin(i); <br />
+ * Camera camera; <br />
+ * while ((camera = system.next()) != null) { <br />
+ * passBatches.get(i).begin(camera); <br />
+ * passBatches.get(i).render(instances, environment); <br />
+ * passBatches.get(i).end(); <br />
+ * } <br />
+ * camera = null; <br />
+ * system.end(i); <br />
+ * } <br />
```

Comment:
This looks a bit (from the API view) as if a `ShadowPass` (or something like that) could be used:
`interface ShadowSystem<T extends ShadowPass> implements Iterable<T>`
So for the user can do something like:
```java
for (ShadowPass pass : system) {
    pass.begin();
    for (Camera cam : pass) {
        pass.getBatch().begin(cam);
        pass.getBatch().render(...);
        pass.getBatch().end();
    }
    pass.end();
}
```
Or something like that. I might be wrong though. E.g. this would make the system aware of the modelbatch and I'm not sure if that fits your design.